src/create.cpp

```diff
@@ -0,0 +1,216 @@
+/// @returns A unique temporary path
+fs::path make_path(std::string_view label, std::string_view extension) {
+  constexpr static std::size_t CHARS_IN_GUID = 39;
+  GUID guid;
+  CoCreateGuid(&guid);
+
+  wchar_t name[CHARS_IN_GUID];
+  swprintf(name, CHARS_IN_GUID,
+           L"%08X-%04X-%04X-%02X%02X-%02X%02X%02X%02X%02X%02X", guid.Data1,
+           guid.Data2, guid.Data3, guid.Data4[0], guid.Data4[1], guid.Data4[2],
+           guid.Data4[3], guid.Data4[4], guid.Data4[5], guid.Data4[6],
+           guid.Data4[7]);
+
+  fs::path pattern = fs::temp_directory_path() / label / name;
+  pattern += extension;
+
+  return pattern;
+}
+#else
+/// Placeholder in temporary path templates to be replaced
+/// with random characters
+constexpr std::string_view placeholder = "XXXXXX";
+
+/// Creates a temporary path pattern with the given label and extension
+/// @note label and extension must be valid
+/// @param[in] label     A label to attach to the path pattern
+/// @param[in] extension An extension of the temporary file path
+/// @returns A path pattern for the unique temporary path
+fs::path make_pattern(std::string_view label, std::string_view extension) {
+  fs::path pattern = fs::temp_directory_path() / label / placeholder;
+  pattern += extension;
+
+  return pattern;
+}
+#endif
+}    // namespace
+
+bool create_parent(const fs::path& path, std::error_code& ec) {
+  return fs::create_directories(path.parent_path(), ec);
+}
+
+std::pair<fs::path, entry::native_handle_type>
+create_file(std::string_view label, std::string_view extension) {
+  validate_label(label);    // throws std::invalid_argument with a proper text
+  validate_extension(extension);
+
+  std::error_code ec;
+  auto file = create_file(label, extension, ec);
+
+  if (ec) {
+    throw fs::filesystem_error("Cannot create a temporary file", ec);
+  }
+
+  return file;
+}
+
+std::pair<fs::path, entry::native_handle_type>
+create_file(std::string_view label, std::string_view extension,
+            std::error_code& ec) {
+  if (!is_label_valid(label) || !is_extension_valid(extension)) {
+    ec = std::make_error_code(std::errc::invalid_argument);
+    return std::pair<fs::path, entry::native_handle_type>();
+  }
+
+#ifdef _WIN32
+  fs::path::string_type path = make_path(label, extension);
+#else
+  fs::path::string_type path = make_pattern(label, extension);
+#endif
+  create_parent(path, ec);
+  if (ec) {
+    return std::pair<fs::path, entry::native_handle_type>();
+  }
+
+#ifdef _WIN32
+  HANDLE handle =
+      CreateFile(path.c_str(), GENERIC_READ | GENERIC_WRITE,
+                 FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE,
+                 nullptr, CREATE_NEW, FILE_ATTRIBUTE_NORMAL, nullptr);
+  if (handle == INVALID_HANDLE_VALUE) {
+    ec = std::error_code(GetLastError(), std::system_category());
+    return std::pair<fs::path, entry::native_handle_type>();
+  }
+#else
+  int handle = mkstemps(path.data(), static_cast<int>(extension.size()));
```

Comment:
An example of mktemp implementation if I ever want to implement it here:

https://github.com/apple-oss-distributions/Libc/blob/af11da5ca9d527ea2f48bb7efbd0f0f2a4ea4812/stdio/FreeBSD/mktemp.c#L66